### ISSUE TITLE
Clarify requirements for the `std::size_t` and `std::ptrdiff_t` type aliases

### ISSUE THREAD

**Issue by tahonermann** [KhronosGroup/SYCL-Docs#373]
The SYCL specification relies on the `std::size_t` and `std::ptrdiff_t` type aliases, but does not currently impose requirements that they be defined on the host and device to either the same type, or to types with matching size and alignment. Can objects of this type transit the host and device boundary or be accessed via USM? Since these types are used in the signatures of standard library functions, their exact definition has ABI implications.

This issue has been filed to update the SYCL specification to more explicitly state the intended requirements for these type aliases.

**Comment by nliber** [comment]
They have to be the same `typedef` or it is an ODR violation, isn't it?  Please see my comments in #371.

**Comment by tahonermann** [comment]
It isn't clear to me how to apply the ODR to SYCL (I'm new to SYCL, so I haven't given this much thought yet). Since SYCL is intended to be a single-source language, almost any realistic use of `__SYCL_DEVICE_ONLY__` or `__SYCL_SINGLE_SOURCE__` seems like it would create ODR concerns if not limited to host-only or device-only code. There is clearly a need to differentiate host and device code at times but I'm not sure the current SYCL specification draws a clear line of demarcation. But that is a much larger concern than this issue seeks to resolve :)

**Comment by nliber** [comment]
The macros (especially `__SYCL_DEVICE_ONLY__`, as it changes value during compilation) can cause ODR violations as well, and is a much harder problem to solve.

Back to the concern at hand.  Suppose both `long` and `long long` are 64-bits, `ptrdiff_t` is a `long` on the host and a `long long` on the device.

What should `std::is_same_v<ptrdiff_t, long>` return?  Beats me.  This isn't really executed anywhere and there are contexts (such as in a `static_assert`, SFINAE condition, etc.) where asking if that is in device code or host code doesn't even make sense.

These are just the simplest questions that have to be answered if one allows them to be different types.

**Comment by gmlueck** [comment]
I agree with @nliber.  We should change the SYCL spec to say that all types which are available in device code must have the same size, alignment requirements, and representation on both host and device.  We say some of this in the spec already, but we don't say all those things, and we don't say it clearly.

I think this is what the committee intended because there is no description in the spec at all about any requirement for the application to convert data formats when passing data to / from kernels.  If we did intend to allow different data type representation between host and device, this would be a pretty significant hole in the spec.  I also agree with @nliber's concern about the ambiguity of type traits if we did allow different representations in host and device.

Given that SYCL is a single-source language, it makes sense that the type system should be consistent across all parts of a source file.
